### DISCUSSION TITLE
Prior art: near-duplicate detection is a cascade problem, not an embedding problem

### DISCUSSION THREAD

**Discussion by dezeat** [dezeat/golearn#137] [Ideas]
**Prior-art briefing, exploratory — not a decision.** Context for #124 after A-22/A-24 showed embedding cosine cannot carry the near-duplicate gate. Per the `scout` guardrails this is evidence, not a verdict; anything that ends up *deciding* something graduates into `DECISIONS.md` with these citations.

## The problem has a name

**Duplicate Question Detection**, classic benchmark **Quora Question Pairs**. There is even an MCQ-specific paper — [*Detecting Duplicate Multiple Choice Questions in the Large Question Bank*](https://dl.acm.org/doi/10.1007/978-3-031-41456-5_17) (Springer 2023; paywalled, method not read).

## Our measured failure is the textbook failure mode

The distinction is **bi-encoder vs cross-encoder**. A bi-encoder embeds each text separately and compares points; a cross-encoder puts both texts through the model together.

Reranking with a cross-encoder exists precisely to filter "context that is superficially similar but semantically irrelevant — a common failure mode of Bi-Encoders" ([Weaviate](https://weaviate.io/blog/cross-encoders-as-reranker), [VeloDB](https://www.velodb.io/glossary/bi-encoder-vs-cross-encoder)). That is exactly our *"zero value of a slice"* vs *"of a map"*.

Sentence-Transformers recommends for Quora dedup specifically: **cross-encoder after bi-encoder pre-filtering** ([sbert.net](https://sbert.net/examples/cross_encoder/training/quora_duplicate_questions/README.html)). Cross-encoders cannot search a corpus — "clustering 10,000 sentences would require ~50 million combinations, about 65 hours" ([sbert.net](https://sbert.net/examples/cross_encoder/applications/README.html)).

## So the accepted answer is a cascade, not a replacement

```
embedding (cheap, wide recall) → top-N candidates → joint-encoding judge (precise)
```

Our embeddings are not the wrong tool — they are the wrong **last** tool. As a recall filter they need only a generous cut, so **no calibrated threshold is required** and D-022 becomes moot rather than evaded.

## Where it collides with our constraints

- **Ollama has no `/api/rerank`.** [PR ollama/ollama#7219](https://github.com/ollama/ollama/pull/7219) has been idle ~a year. A classic cross-encoder is not reachable through our transport.
- **Qwen3-Reranker is a CausalLM, not a classifier** — it emits the probability of `yes`/`no` tokens ([HF card](https://huggingface.co/Qwen/Qwen3-Reranker-0.6B), [vLLM](https://docs.vllm.ai/en/v0.9.2/examples/offline_inference/qwen3_reranker.html)), so it runs over `/api/chat`. But Ollama returns text, not logits: we would get a binary verdict without a graded score, and the ladder uses the score to route repair vs replace.
- **FTS5 cannot be the recall filter** (the cheap option D-021 named). It is lexical, so it would miss precisely the disjoint-vocabulary duplicates the judge exists to catch.

## LLM-as-judge: known pitfalls to price in

1. **Position bias is real and documented.** Pairwise verdicts flip when the order flips; the standard mitigation is running both orders and aggregating, with *position consistency* and *preference fairness* as metrics ([arXiv 2406.07791](https://arxiv.org/abs/2406.07791), [ACL 2025](https://aclanthology.org/2025.ijcnlp-long.18/)). Costs double the calls.
2. **The threshold does not disappear, it becomes invisible.** The decision boundary moves from a committed, versioned number into weights and a prompt. Regarding D-022 that is a *loss* of inspectability unless the judge is measured against the same fixture set — which is what A-25 does.
3. **Determinism.** `Gate.Apply` promises in-code that "the same pack screened twice must resolve the same way". A judge breaks that promise.

**The neat fit:** our fixture set already labels in the taxonomy `identical / paraphrase / semantic / competency / concept / unrelated`. A judge asked for exactly that label maps onto the ladder rungs, and the 13 pairs become an **accuracy benchmark** instead of a calibration set — same evidence, no threshold.

## LangChain / langchaingo: no

- Python; the Go port [`tmc/langchaingo`](https://github.com/tmc/langchaingo/) is mid [community-maintenance handover](https://github.com/sayerxofficial/langchaingo) and already forked.
- It would duplicate what exists: `ports.Provider` + four adapters *is* that abstraction. Chains/memory/retrievers are not needed.
- "Multiple LLM calls" is a loop, not a framework. What is actually needed is a prompt template and a parser — double-digit lines against the existing port.

## Unpriced

At the N100 host's ~30 s per judgement, a naive gate is ~100 minutes per pack. Batching neighbours into one call per candidate is the obvious lever and is unmeasured.
